ddl/table.go

```diff
@@ -147,6 +148,81 @@ func onDropTableOrView(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	return ver, errors.Trace(err)
 }
 
+func (w *worker) onRestoreTable(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	// check gc enable status again.
+	gcEnable, err := checkGCEnable(w)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	if gcEnable {
+		job.State = model.JobStateCancelled
+		return ver, errors.Errorf("can not restore deleted table when gc is enable")
+	}
+
+	schemaID := job.SchemaID
+	tbInfo := &model.TableInfo{}
+	var autoID, dropJobID int64
+	var enableGCAfterRecover bool
+	if err = job.DecodeArgs(tbInfo, &autoID, &dropJobID, &enableGCAfterRecover); err != nil {
+		// Invalid arguments, cancel this job.
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	err = checkTableNotExists(t, job, schemaID, tbInfo.Name.L)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	ver, err = updateSchemaVersion(t, job)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	// Remove dropped table DDL job from gc_delete_range table.
+	err = w.delRangeManager.removeFromGCDeleteRange(dropJobID, tbInfo.ID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	tbInfo.State = model.StatePublic
+	tbInfo.UpdateTS = t.StartTS
+	err = t.CreateTableAndSetAutoID(schemaID, tbInfo, autoID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	if enableGCAfterRecover {
+		err = enableGC(w)
```

Comment:
If you finish enableGC, but update the job transaction is failed, the job will retry, and the gc is enabled, and the job will be canceled in line 159.

Reply:
You should do enabling it in `finishDDLJob`

Reply:
So please use gofail to mock an error in line 202.

Reply:
can not mock error in line 202, It is impossible error here.

Reply:
I add gofail to mock commit error in tikv/txn.go, PTAL.